qubesadmin/tools/qvm_start_daemon.py

```diff
@@ -709,59 +828,56 @@ def is_watched(self, vm):
 
-parser = qubesadmin.tools.QubesArgumentParser(
-    description='start GUI for qube(s)', vmname_nargs='*')
-parser.add_argument('--watch', action='store_true',
-                    help='Keep watching for further domain startups')
-parser.add_argument('--force-stubdomain', action='store_true',
-                    help='Start GUI to stubdomain-emulated VGA,'
-                         ' even if gui-agent is running in the VM')
-parser.add_argument('--pidfile', action='store', default=pidfile_path,
-                    help='Pidfile path to create in --watch mode')
-parser.add_argument('--notify-monitor-layout', action='store_true',
-                    help='Notify running instance in --watch mode'
-                         ' about changed monitor layout')
-parser.add_argument('--kde', action='store_true',
-                    help='Set KDE specific arguments to gui-daemon.')
-# Add it for the help only
-parser.add_argument('--force', action='store_true', default=False,
-                    help='Force running daemon without enabled services'
-                         ' \'guivm\' or \'audiovm\'')
+    only_if_service_enabled = ['guivm', 'audiovm']
+    enabled_services = [
+        service for service in only_if_service_enabled
+        if os.path.exists(f'/var/run/qubes-service/{service}')
+    ]
 
+    if "--force" in sys.argv or os.path.exists("/etc/qubes-release"):
+        enabled_services = only_if_service_enabled
 
-def main(args=None):
-    """ Main function of qvm-start-daemon tool"""
-    only_if_service_enabled = ['guivm', 'audiovm']
-    enabled_services = [service for service in only_if_service_enabled if
-                        os.path.exists('/var/run/qubes-service/%s' % service)]
-    if not enabled_services and '--force' not in sys.argv and \
-            not os.path.exists('/etc/qubes-release'):
-        print(parser.format_help())
+    if not enabled_services:
+        log.info("None of 'audiovm' nor 'guivm' service is enabled")
+        log.info(parser.format_help())
         return
-    args = parser.parse_args(args)
+
+    args = parser.parse_args()
+
+    if args.debug:
+        log.setLevel(logging.DEBUG)
+
     if args.watch and args.notify_monitor_layout:
         parser.error('--watch cannot be used with --notify-monitor-layout')
 
     if args.all_domains:
         vm_names = None
     else:
         vm_names = [vm.name for vm in args.domains]
+
     launcher = DAEMONLauncher(
-        args.app,
+        app=args.app,
+        enabled_services=enabled_services,
         vm_names=vm_names,
-        kde=args.kde)
+        kde=args.kde
+    )
 
     if args.watch:
-        fd = os.open(args.pidfile,
-                     os.O_RDWR | os.O_CREAT | os.O_CLOEXEC,
-                     0o600)
+        # Start qubes running before we started qvm-start-daemon
+        for qube in launcher.app.domains:
+            if qube.name == "dom0":
+                continue
+            launcher.on_connection_established(
```

Comment:
This is wrong, it should be called only when really receiving `connection-established` event, otherwise you will miss anything that happens before you start receiving events.